datafusion/expr/src/type_coercion/functions.rs

```diff
@@ -130,6 +130,31 @@ fn get_valid_types(
             _ => Ok(vec![vec![]]),
         }
     }
+    fn array_append_and_optional_index(
+        current_types: &[DataType],
+    ) -> Result<Vec<Vec<DataType>>> {
+        // make sure there's at least 2 arguments
+        if !(current_types.len() == 2 || current_types.len() == 3) {
+            return Ok(vec![vec![]]);
+        }
+
+        let first_two_types = &current_types[0..2];
+        let valid_types = array_append_or_prepend_valid_types(first_two_types, true)?;
+
+        let valid_types_with_index = valid_types
```

Comment:
I think we can optionally append valid_types_with_index if the len is 3, otherwise just return here.

Reply:
Thanks! Addressed in [5086f31](https://github.com/apache/arrow-datafusion/pull/9233/commits/5086f3150b3533f7bc7bb3e60d8561a24b5f6797).